mitxgraders/comparers/linear_comparer.py

```diff
@@ -0,0 +1,169 @@
+    """
+    Get total error in the difference between two samples.
+    Arguments:
+        x, y: compatible numpy arrays
+    """
+    return np.sqrt(sum(np.square(x - y)))
+
+class LinearComparer(CorrelatedComparer):
+    """
+    Used to check that there is an linear relationship between student's input
+    and the expected answer.
+
+    The general linear relationship is expected = a * student + b. The comparer
+    can check for four subtypes:
+        equals: (a, b) = (1, 0)
+        proportional: b = 0
+        offset: a = 1
+        linear: neither a nor b fixed
+
+    Configuration
+    =============
+    The first four configuration keys determine the amount of partial credit
+    given for a specific type of linear relationship. If set to None, the
+    relationship is not checked.
+        equals (None | number): defaults to 1.0
+        proportional (None | number): defaults to 0.5
+        offset (None | number): defaults to None
+        linear (None | number): defaults to None
+
+    The remaining configuration keys specify a feedback message to be given
+    in each case:
+        equals_msg (str): defaults to ''
+        proportional_msg (str): defaults to 'The submitted answer differs from
+            an expected answer by a constant factor.'
+        offset_msg (str): defaults to ''
+        linear_msg (str): defaults to ''
+
+    NOTE:
+    """
+
+    schema_config = Schema({
+        Required('equals', default=1.0): Any(None, Range(0, 1)),
+        Required('proportional', default=0.5): Any(None, Range(0, 1)),
+        Required('offset', default=None): Any(None, Range(0, 1)),
+        Required('linear', default=None): Any(None, Range(0, 1)),
+        Required('equals_msg', default=''): str,
+        Required('proportional_msg', default=(
+            'The submitted answer differs from an expected answer by a '
+            'constant factor.'
+        )): str,
+        Required('offset_msg', default=''): str,
+        Required('linear_msg', default=''): str,
+    })
+
+    def __init__(self, config=None, **kwargs):
+        super(LinearComparer, self).__init__(config, **kwargs)
+        all_modes = ('equals', 'proportional', 'offset', 'linear')
+        self.modes = tuple(mode for mode in all_modes if self.config[mode] is not None)
+
+    error_calculators = {
+        'equals': get_equals_fit_error,
+        'proportional': get_proportional_fit_error,
+        'offset': get_offset_fit_error,
+        'linear': get_linear_fit_error,
+    }
+
+    def __call__(self, comparer_params_evals, student_evals, utils):
+        student_eval_norm = np.linalg.norm(student_evals)
+
+        # Validate student input shape...only needed for MatrixGrader
+        try:
+            utils.validate_shape(student_evals[0], comparer_params_evals[0][0].shape)
+        except AttributeError:
+            pass # not called by MatrixGrader
+
+        if is_nearly_zero(student_eval_norm, utils.tolerance, reference=comparer_params_evals):
```

Comment:
Uhh... the reference here is off. `comparer_params_evals` is a list of evaluations. Also, `student_eval_norm` is a single number; shouldn't this comparison be done for each input individually? Finally, I think putting this statement here precludes checking for equality when the answer is actually 0.